docs/design/proposals/creds-from-kms.md

```diff
@@ -0,0 +1,130 @@
+# Fetching Ceph user credentials from KMS
+
+## Proposal
+
+Ceph-CSI supports several KMS implementations for storing and
+fetching DEKs (Data Encryption Keys) used for RBD volume encryption.
+The focus of this proposal is to leverage existing KMS implementation
+to support obtaining Ceph user keyring used while
+creating, deleting, mapping and resizing RBD volumes.
+
+### Benefits
+
+- For scenarios where Ceph is deployed externally, either as
+  a standalone or through Rook, one needs to obtain the Ceph keyring(s)
+  and manually create the K8s secrets needed by Ceph-CSI. With this feature,
+  sensitive keyring(s) can be stored securely in an
+  external secret management system such as Hashicorp Vault,
+  AWS Secret Manager, etc., and have Ceph-CSI pull these
+  keys directly from them.
+- More secure compared to K8s secrets,
+  which are relatively easy to access and
+  store sensitive secrets as unencrypted based64 encoded text.
+
+### Drawbacks
+
+- Adds additional overhead as on each CSI RPC call
+  a request is made to KMS for fetching credentials.
+- Possible risk of hitting the KMS rate limit
```

Comment:
Wont this be a major drawback?

Reply:
I feel KMS rate limit is a valid concern for anything, including RBD volume encryption, using the KMS implementation. One possible approach to mitigate this would be to use an in-memory cache of Ceph keyrings, given that keyrings don't change too often we only fetch them when the keyring doesn't exist already or is invalid (i.e. on failed ceph auth).

Reply:
:+1: the in-memory cache need to be deleted for few cases like 
* auth issues
* connection issues

Reply:
I added a point on in-memory cache under `Code Changes` sections. 

Do you mind elaborating a bit on what you mean by `connection issues` in this context?